auth_token/utils.py

```diff
@@ -176,21 +312,333 @@ def get_user(request):
+    Args:
+        authorization_request: authorization request where new OTP will be created.
+        otp_key_generator: function which generates OTP.
+        expiration: expiration time in seconds. Empty value means that original expiration time will be used.
+
+    Returns:
+        AuthorizationRequest instance with new OTP
+    """
+    assert authorization_request.type == AuthorizationRequestType.OTP
+    assert authorization_request.result is None
+
+    if expiration is None:
+        expiration = (authorization_request.expires_at - authorization_request.created_at).total_seconds()
+
+    authorization_request.change_and_save(expires_at=compute_expires_at(expiration))
+    otp = create_otp(authorization_request.slug, key_generator=otp_key_generator, expiration=expiration)
+    authorization_request.one_time_passwords.all().change_and_save(is_active=False)
+    authorization_request.one_time_passwords.add(otp)
+    authorization_request.secret_key = otp.secret_key
+    return authorization_request
+
+
+def _get_valid_mobile_device(mobile_device_id, mobile_login_token):
+    """
+    Retrun mobile device with valid ID and token.
+    Args:
+        mobile_device_id: mobile device ID.
+        mobile_login_token: mobil device secret token.
+
+    Returns:
+        valid mobile device or None.
     """
-    return force_text(binascii.hexlify(os.urandom(length)))
+    from .models import MobileDevice
+
+    for mobile_device in MobileDevice.objects.filter(uuid=mobile_device_id, is_active=True):
+        if mobile_device.check_password(mobile_login_token):
+            return mobile_device
+
+    return None
+
+
+def check_authorization_request(authorization_request, mobile_device_id=None, mobile_login_token=None,
+                                otp_secret_key=None):
+    """
+    Check if data send by client is valid with autorization request and access can be granted.
+    Args:
+        authorization_request: validated authorization request
+        mobile_device_id: mobile device ID which can verify authorization request. Only valid for DEVICE type.
+        mobile_login_token: mobile device token which can verify authorization request. Only valid for DEVICE type.
+        otp_secret_key: OTP code which can verify authorization request. Only valid for OTP type.
+
+    Returns:
+        True for valid input data, False otherwise.
+    """
+    if authorization_request.state != AuthorizationRequestState.WAITING:
+        return False
+
+    if authorization_request.type == AuthorizationRequestType.OTP:
+        return bool(otp_secret_key) and (
+            any(
+                check_otp(otp.slug, otp_secret_key) for otp in authorization_request.one_time_passwords.filter(
+                    is_active=True
+                )
+            ) or (
+                    settings.AUTHORIZATION_REQUEST_OTP_DEBUG_CODE
+                    and settings.AUTHORIZATION_REQUEST_OTP_DEBUG_CODE == otp_secret_key
+            )
+        )
+    else:
```

Comment:
This is quite limiting if another auth request type will be added :( Better for future extension would be if each auth request type would have dedicated checker class. But never mind.

Reply:
Or some kind of backend.